scipy/stats/stats.py

```diff
@@ -4153,6 +4154,178 @@ def binary_search(n, n1, n2, side):
     return oddsratio, pvalue
 
 
+def barnard_exact(
+        table,
+        alternative='two-sided',
+        variance="pooled",
+        nuisance_nbre=500):
+    """
+    Perform a Barnard exact test on a 2x2 contingency table.
+
+    Parameters
+    ----------
+    table : array_like of ints
+        A 2x2 contingency table.  Elements should be non-negative integers.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+          * 'two-sided'
+          * 'less': one-sided
+          * 'greater': one-sided
+
+    variance : {'pooled', 'unpooled'}, optional
+        Defines how to calcul the score statistic.
+        The following options are available (default is 'pooled'):
+
+          * 'pooled'
+          * 'unpooled'
+    nuisance_nbre : int, optional
+        Number of nuisance's samples to generate. Default is 1000. Must be 
+        non-negative
+
+    Returns
+    -------
+    score_statistic : float
+        This is prior odds ratio and not a posterior estimate.
+    p_value : float
+        P-value, the probability of obtaining a distribution at least as
+        extreme as the one that was actually observed, assuming that the
+        null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    Examples
+    --------
+    examples from Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials.
+    (https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf)
+
+    Consider the following example of a vaccine efficacy study (Chan, 1998). 
+    In a randomized clinical trial of 30 subjects, 15 were innoculated with a
+    recombinant DNA influenza vaccine and the 15 were innoculated with a
+    placebo. Twelve of the 15 subjects in the placebo group (80%) eventually
+    became infected with influenza whereas for the vaccine group, only 7 of 
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2 
+    table :
+
+                    Vaccine  Placebo
+        Yes             8        2
+        No              1        5
+
+    We use this table to find the p-value:
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
```

Comment:
Here (and below), I'm thorn; variable that start with upper case are usually classes, and full upper case constants. We can leave it as is and see how scipy maintainer react.